File: day19.py
from typing import NamedTuple

class Materials(NamedTuple):
    ore: int = 0
    clay: int = 0
    obsidian: int = 0
    geode: int = 0

    def __add__(self, other):
        return Materials(*(a+b for (a,b) in zip(self, other)))

    def __sub__(self, other):
        return Materials(*(a-b for (a,b) in zip(self, other)))

    def __mul__(self, m):
        return Materials(*(a*m for a in self))

    def __ge__(self, other):
        return all(a >= b for a,b in zip(self, other))

class State(NamedTuple):
    t: int
    mats: Materials
    prod: Materials

def parse(line):
    words = line.split()
    return {
        Materials(geode=1): Materials(ore=int(words[27]), obsidian=int(words[30])),
        Materials(obsidian=1): Materials(ore=int(words[18]), clay=int(words[21])),
        Materials(clay=1): Materials(ore=int(words[12])),
        Materials(ore=1): Materials(ore=int(words[6])),
    }

def run(bp, init:State, N):
    max_c = tuple(max(v[i] for v in bp.values()) for i in range(4))

    visited = set()
    queue = [init]
    while queue:
        s = queue.pop()
        if s.t == N:
            yield s.mats.geode
            continue
        for p, c in bp.items():
            x = s.prod+p
            if x[0] > max_c[0] or x[1] > max_c[1] or x[2] > max_c[2]:
                continue

            new_s = s
            while new_s.t < N:
                if new_s.mats >= c: 
                    new_s = State(new_s.t+1, new_s.mats + new_s.prod - c, new_s.prod + p)

                    if new_s not in visited:
                        visited.add(new_s)
                        queue.append(new_s)
                    break
                new_s = State(new_s.t+1, new_s.mats + new_s.prod, new_s.prod)
    print("visited", len(visited))

data = open("data/day19").read()

blueprints = [parse(line) for line in data.split("\n")]

s = State(0, Materials(), Materials(ore=1))

total = 0
for id, bp in enumerate(blueprints, 1):
    geodes = max(run(bp, s, 24))
    print(id, geodes, "->", geodes*id)
    total += geodes*id
print(x := total, x == 2341)

total = 1
for id, bp in enumerate(blueprints[:3], 1):
    geodes = max(run(bp, s, 32))
    print(id, geodes)
    total *= geodes
print(x := total, x == 3689)

File: src/day19.rs
use std::{ops::{Add, Sub, Mul}, collections::BTreeSet, cmp::max};
use regex::Regex;

#[derive(Debug, Copy, Clone, Eq, PartialEq, PartialOrd, Ord)]
struct Mats(u8,u8,u8,u8);

impl Mats {
    fn ge(self, rhs: Self) -> bool {
        self.0 >= rhs.0 && self.1 >= rhs.1 && self.2 >= rhs.2 && self.3 >= rhs.3
    }
}

impl Add for Mats {
    type Output=Mats;

    fn add(self, rhs: Self) -> Self::Output {
        Mats(self.0+rhs.0, self.1+rhs.1, self.2+rhs.2, self.3+rhs.3)
    }
}

impl Sub for Mats {
    type Output=Mats;

    fn sub(self, rhs: Self) -> Self::Output {
        Mats(self.0-rhs.0, self.1-rhs.1, self.2-rhs.2, self.3-rhs.3)
    }
}

impl Mul<u8> for Mats {
    type Output=Mats;

    fn mul(self, rhs: u8) -> Self::Output {
        Mats(self.0*rhs, self.1*rhs, self.2*rhs, self.3*rhs)
    }
}

#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug)]
struct State {
    t: u8,
    res: Mats,
    prod: Mats,
}

fn parse(line: &str) -> Option<[(Mats, Mats); 4]> {
    let re = Regex::new(r"^Blueprint \d+: Each ore robot costs (\d+) ore. Each clay robot costs (\d+) ore. Each obsidian robot costs (\d+) ore and (\d+) clay. Each geode robot costs (\d+) ore and (\d+) obsidian.$").ok()?;
    let c = re.captures(line)?;

    Some([
        (Mats(1,0,0,0), Mats(c.get(1)?.as_str().parse().ok()?,0,0,0)),
        (Mats(0,1,0,0), Mats(c.get(2)?.as_str().parse().ok()?,0,0,0)),
        (Mats(0,0,1,0), Mats(c.get(3)?.as_str().parse().ok()?,c.get(4)?.as_str().parse().ok()?,0,0)),
        (Mats(0,0,0,1), Mats(c.get(5)?.as_str().parse().ok()?,0,c.get(6)?.as_str().parse().ok()?,0)),
    ])
}

fn run(bp: [(Mats, Mats);4], n: u8) -> u8 {
    let mut max_c = Mats(0,0,0,0);
    for (_, cost) in bp {
        max_c.0 = max(max_c.0, cost.0);
        max_c.1 = max(max_c.1, cost.1);
        max_c.2 = max(max_c.2, cost.2);
    }

    let mut visited: BTreeSet<State> = BTreeSet::new();
    let mut queue = vec![
        State{ t: 0, res: Mats(0,0,0,0), prod: Mats(1,0,0,0) }
    ];

    let mut best = 0;

    while let Some(s) = queue.pop() {
        if s.t == n {
            if s.res.3 > best {
                best = s.res.3;
            }
            continue
        }

        for &(prod, cost) in bp.iter() {
            let x = s.prod + prod;
            if x.0 > max_c.0 || x.1 > max_c.1 || x.2 > max_c.2 {
                continue
            }

            let mut new_s = s;
            while new_s.t < n {
                if new_s.res.ge(cost) {
                    new_s = State{
                        t: new_s.t + 1,
                        res: new_s.res + new_s.prod - cost,
                        prod: new_s.prod + prod,
                    };
                    if visited.insert(new_s) {
                        queue.push(new_s)
                    }
                    break
                }
                new_s = State{
                    t: new_s.t + 1,
                    res: new_s.res + new_s.prod,
                    prod: new_s.prod,
                };
            }
        }
    }
    println!("visited {}", visited.len());
    best
}


pub fn solve(data: &str) -> Option<(u16, u16)> {
    let blueprints = data.split('\n').map(parse).collect::<Option<Vec<_>>>()?;

    let part1 = blueprints.iter().enumerate().map(|(t, &bp)| {
        (t+1) as u16 * run(bp, 24) as u16
    }).sum();

    let part2 = run(blueprints[0], 32) as u16 * run(blueprints[1], 32) as u16 * run(blueprints[2], 32) as u16;

    return Some((part1, part2))
}